crates/services/p2p/src/p2p_service.rs

```diff
@@ -217,29 +222,46 @@ impl FuelP2PService {
         let tcp_config = tcp::Config::new().port_reuse(true);
         let behaviour = FuelBehaviour::new(&config, codec.clone())?;
 
-        let mut swarm = SwarmBuilder::with_existing_identity(config.keypair.clone())
+        let swarm_builder = SwarmBuilder::with_existing_identity(config.keypair.clone())
             .with_tokio()
             .with_tcp(
                 tcp_config,
                 transport_function,
                 libp2p::yamux::Config::default,
             )
-            .map_err(|_| anyhow::anyhow!("Failed to build Swarm"))?
-            .with_dns()?
-            .with_behaviour(|_| behaviour)?
-            .with_swarm_config(|cfg| {
-                if let Some(timeout) = config.connection_idle_timeout {
-                    cfg.with_idle_connection_timeout(timeout)
-                } else {
-                    cfg
-                }
-            })
-            .build();
+            .map_err(|_| anyhow::anyhow!("Failed to build Swarm"))?;
+
+        let mut swarm = if metrics {
+            // we use the global registry to store the metrics without needing to create a new one
+            // since libp2p already creates sub-registries
+            let mut registry = global_registry().registry.lock();
+
+            swarm_builder
+                .with_bandwidth_metrics(&mut registry)
+                .with_behaviour(|_| behaviour)?
+                .with_swarm_config(|cfg| {
+                    if let Some(timeout) = config.connection_idle_timeout {
+                        cfg.with_idle_connection_timeout(timeout)
+                    } else {
+                        cfg
+                    }
+                })
+                .build()
+        } else {
+            swarm_builder
+                .with_behaviour(|_| behaviour)?
+                .with_swarm_config(|cfg| {
+                    if let Some(timeout) = config.connection_idle_timeout {
+                        cfg.with_idle_connection_timeout(timeout)
+                    } else {
+                        cfg
+                    }
+                })
+                .build()
+        };
```

Comment:
libp2p generics prevent me from doing something like this -
```rs
let swarm_builder = initial_stuff;
let swarm builder = if metrics {
  let mut registry = ..;
  swarm_builder.with_bandwidth_metrics(&mut registry)
} else {
  swarm_builder
}
.. rest of the stuff with_behavior etc
```

Reply:
yep, I have tried the same. There is a function `with_no_bandwidth_metrics()` that could help here, but it is implemented for a different phase that can be reached only via private functions :(